### ISSUE TITLE
Extra latency in the first `for` iteration

### ISSUE THREAD

**Issue by Unike267** [stnolting/neorv32#888]
Hi @stnolting !

I'm performing a CFS (Custom Function Subsystem) latency/throughput measurement. 

For this purpose I've added a multiplier with NEORV32 through this mode of connection.

In this context I've made the following code in C:

``` c
  // Declaration of variables 
  int i;
  //0000000000000001 x 0000000000000001
  static uint32_t fir = 0x00010001;
  //0000000000000010 x 0000000000000010
  static uint32_t sec = 0x00020002;
  //0000000000000100 x 0000000000000100
  static uint32_t thi = 0x00040004;
  //0000000000001000 x 0000000000001000
  static uint32_t fou = 0x00080008;

 // Intro
  neorv32_uart0_printf("\n CFS-lat \n\n");
  // Write 4 inputs to mult and read the outputs from mult one by one
    for (i=0; i<4 ; i++) {
    if(i==0){
            NEORV32_CFS->REG[0] = fir; // Write fir to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fir to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fir from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
           } 
    if(i==1){
            NEORV32_CFS->REG[0] = sec; // Write sec to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write sec to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read sec from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
           }   
    if(i==2){
            NEORV32_CFS->REG[0] = thi; // Write thi to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write thi to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read thi from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
           }   
    if(i==3){
            NEORV32_CFS->REG[0] = fou; // Write fou to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fou to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fou from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
           }   
    } 
  // End

``` 

I've made a test bench with VUnit and I've captured the `bus_req` and the `bus_rsp` through external names (see #886 ) to evaluate the write/read time and perform the latency measurement.  

The question is that in the first `for` loop iteration extra latency is added, specifically 2 cycles after REG[0] addressing and another 2 cycles  after REG[1] addressing, 4 cycles of extra latency in total. As we can see in the following image:

![issue_cfs_1](https://github.com/stnolting/neorv32/assets/36209778/0ecf9312-64f3-4eba-bae9-72b887b1bb4e)

That is, the normal CFS latency for this test is 20 cycles, but for the first `for` iteration it is 24 cycles.

A closer look:

![issue_cfs_4](https://github.com/stnolting/neorv32/assets/36209778/c39bc1dc-cd71-43cd-9724-520a5fda1b14)

Here's a closer look at normal operation (specifically in the second `for` iteration `i==1`):

![issue_cfs_2](https://github.com/stnolting/neorv32/assets/36209778/32c9ad7c-235f-48bc-9848-e48c534ee7e5)

*Note: in the previous images the `BUS_REQ` and the `BUS_RSP` signals are from NEORV32 and the `FIFO_IN`, `FIFO_OUT` and `FULL/EMPTY` signals are from multiplier.*

The question is: why is this happening?

I have seen that what follows the REG[0] and REG[1] addressing is a counter that increments by 4. But why does it increase twice in the first iteration? 

I've tried to initialize the registers because maybe these extra cycles could be for write the first value in the memory-mapped registers but that is not. 😅

I look forward to your reply,

Cheers! 😃 

---

/cc @umarcor

**Comment by stnolting** [comment]
Hey @Unike267.

> The question is: why is this happening?

That's a good question. 😅
I'm not really sure, but I think this might be an (additional) instruction fetch.

Do you use any caches in your setup?

> Note: in the previous images the BUS_REQ and the BUS_RSP signals are from NEORV32 [...]

What signals are these exactly? The ones from the CFS, right?

Better use the "global" request bus that goes into the GATEWAY as this bus "still contains" all the relevant information to identify each single bus cycle.

https://github.com/stnolting/neorv32/blob/49bc8bee0caa352d4ba7339142ccb9ec9bfcc1df/rtl/core/neorv32_top.vhd#L776

Then, you could add the `src` and `stb` signals from the `bus_req_t` type to see if the current bus access if accessing data or if it is fetching an instruction.

https://github.com/stnolting/neorv32/blob/49bc8bee0caa352d4ba7339142ccb9ec9bfcc1df/rtl/core/neorv32_package.vhd#L127-L137

**Comment by Unike267** [comment]
Hi @stnolting !

> Do you use any caches in your setup?

No, I don't.

> What signals are these exactly? The ones from the CFS, right?

Yes, these are the signals:

https://github.com/stnolting/neorv32/blob/49bc8bee0caa352d4ba7339142ccb9ec9bfcc1df/rtl/core/neorv32_cfs.vhd#L32-L33

> Better use the "global" request bus that goes into the GATEWAY as this bus "still contains" all the relevant information to identify each single bus cycle. 
Then, you could add the src and stb signals from the bus_req_t type to see if the current bus access if accessing data or if it is fetching an instruction.

Okey that's good tip!  😄 I will test it tomorrow when I get to the lab! 🔬

The fact that the CFS `bus_req_i.addr` signal is an ascending counter after puts the address of the memory-mapped register tells you something, maybe something related to the `for loop`  🤔

Anyway, thank you so much for your quickly answer! 🚀

I will keep updating this issue! 😎

Cheers! 😃

**Comment by Unike267** [comment]
Hi @stnolting !

I've relaunched the simulation capturing the "global" request bus (into the GATEWAY) and I've identified that the bus is fetching an instruction when it adds extra latency. As we can see in the following image:

![issue_cfs_5](https://github.com/stnolting/neorv32/assets/36209778/45f98bf2-b60c-44be-8e6d-fdf03b4fde45)

Now I would like to capture the instructions that are executing at those moments, could you tell me what signals I should capture?

I would like to reconstruct 🔧 the `for-loop` to see why extra latency is added and if it can be fixed by software. 

If this is not possible I could add one more iteration to the `for-loop` initializing the registers in the first iteration and start measuring latency from the second iteration 🤔. Something like this:

```c
    for (i=0; i<5 ; i++) {
    if(i==0){
            NEORV32_CFS->REG[0] = 0; // Initialize CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 0; // Initialize CFS memory-mapped register 1
           } 
    if(i==1){
            NEORV32_CFS->REG[0] = fir; // Write fir to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fir to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fir from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
           } 
    if(i==2){
            NEORV32_CFS->REG[0] = sec; // Write sec to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write sec to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read sec from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
           }   
    if(i==3){
            NEORV32_CFS->REG[0] = thi; // Write thi to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write thi to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read thi from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
           }   
    if(i==4){
            NEORV32_CFS->REG[0] = fou; // Write fou to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fou to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fou from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
           }   
    } 
  // End
``` 

But this way we would be left with the doubt. 😅

Anyway thanks so much for the feedback,

Cheers! 😃

**Comment by stnolting** [comment]
> and I've identified that the bus is fetching an instruction when it adds extra latency. As we can see in the following image:

:+1:

> Now I would like to capture the instructions that are executing at those moments, could you tell me what signals I should capture?

This is the instruction word being executed "right now":

https://github.com/stnolting/neorv32/blob/49bc8bee0caa352d4ba7339142ccb9ec9bfcc1df/rtl/core/neorv32_cpu_control.vhd#L166

And this is the according program counter:

https://github.com/stnolting/neorv32/blob/49bc8bee0caa352d4ba7339142ccb9ec9bfcc1df/rtl/core/neorv32_cpu_control.vhd#L171

> I would like to reconstruct 🔧 the for-loop to see why extra latency is added and if it can be fixed by software.

This is not a "bug" or something - it's just the normal way of execution. 😅 Seems like the instruction prefetch buffer went empty at the beginning of the loop (maybe because of a branch) and it needs refilling before the core can proceed.

Changing the C code might also change the precise timing (of instruction/data bus congestion), but if you really need _precise_ execution times I suggest to:
* add the instruction
* make sure the entire loop fits into a cache block
* put your loop code into a function and align that to a cache block

**Comment by Unike267** [comment]
Hi!

I have a little update.

> This is the instruction word being executed:
And this is the according program counter:

I've captured these signals in my test bench. This has not been tribal. External names need that when you reference a type it must be shared through a library. Unlike `bus_req` and `bus_rsp`, `execute_engine_t` type is not defined in `neorv32_package.vhd`. Therefore, I've moved the definition of `execute_engine_t` from `neorv32_cpu_control.vhd` to  `neorv32_package.vhd` and then I've captured `ir` and `pc` signals using the following code:

```vhdl
execute_engine <= << signal .tb_complex_mult_wfifos_cfs.neorv32_mult_wfifos_cfs_0.neorv32_top_inst.core_complex.neorv32_cpu_inst.neorv32_cpu_control_inst.execute_engine : execute_engine_t >>;

instruction <= execute_engine.ir;
pc <= execute_engine.pc;
```
I've obtained the following results:

![issue_cfs_6](https://github.com/stnolting/neorv32/assets/36209778/6ceece0f-e50a-4b5b-b943-1447e7ee4915)

As we can see in the first iteration an extra instruction is executed after addressing REG[0] `0x00100713` and another one after addressing REG[1] `0x00200793`. That is, the sequence of instructions that happens in the latency measurement range is `0xB0E42223` -> `0xB0F42223` -> `0xB0042223` -> `0xB0042683` but in the first iteration two extra instructions are executed.  

I can't do more because I don't know what these instructions mean.  😅 I could deduce that the first part `0xB...` corresponds to the ISA extension maybe the C extension (jumps instructions etc.) but without table I'm stuck.  😢

> This is not a "bug" or something - it's just the normal way of execution. 😅 Seems like the instruction prefetch buffer went empty at the beginning of the loop (maybe because of a branch) and it needs refilling before the core can proceed.

Yeah, I assumed that this was normal cpu operation but the mismatch in the latency measurement gives me a little bit ocd.  😂

Anyway, I'm going to try this:

> Changing the C code might also change the actual timing (of instruction/data bus congestion), but this is hard to anticipate. If you really need more precise (actually, "precise" is the wrong word here...) execution times I suggest to:

>  -  add the instruction cache

> -  make sure the entire loop fits into a cache block

> -   put your loop code into a function and align that to a cache block's size

> Then, you can execute your loop entirely from the I-cache so there won't be any processor bus congestion at all.

Thank you so much for the feedback! 🚀

I will keep updating this issue! 😎

Cheers! 😃

**Comment by stnolting** [comment]
> I've captured these signals in my test bench. This has not been tribal. External names need that when you reference a type it must be shared through a library.

Hmm, I never thought about this. However, I would just use some local "temp" signal that can be accessed more easily:

```vhdl
signal debug_signal : std_ulogic_vector(31 downto 0);
...
debug_signal <= execute_engine.ir;
```

> I can't do more because I don't know what these instructions mean. 😅 I could deduce that the first part 0xB... corresponds to the ISA extension maybe the C extension (jumps instructions etc.) but without table I'm stuck. 😢

You can find the instruction encoding in https://github.com/stnolting/neorv32/blob/main/docs/references/riscv-spec.pdf starting at page 150.

However, it is easier to take the assembly file and use the program counter (`execute_engine .ir`) to identify the specific instructions. Just add the `asm` target to your make command. This will give you something like this (instruction address in the first column):

![grafik](https://github.com/stnolting/neorv32/assets/22944758/d3ad2ae0-7944-4eca-b8a0-933221faa9a6)

**Comment by Unike267** [comment]
Hi @stnolting !

> You can find the instruction encoding in https://github.com/stnolting/neorv32/blob/main/docs/references/riscv-spec.pdf starting at page 150.

Ohh! 😲 I didn't know of the existence of this document, looks very interesting! 

> However, it is easier to take the assembly file and use the program counter (execute_engine .ir) to identify the specific instructions. Just add the asm target to your make command. This will give you something like this (instruction address in the first column):

Okey, thank you so much for the hint! 😄

I'm still working on this issue but I don't want to spend too much time on it so if I don't get the same cycles throughout the loop I'll write something like this:

"The CFS latency measurement for this multiplier is 20 cycles + extra latency*. *4 cycles only in some iterations"

(Maybe written in a more elegant way.)

Anyway thanks for the feedback :rocket: 

Cheers! :smiley:

**Comment by stnolting** [comment]
Could you show your assembly code of that loop? Maybe we can figure out why there is an additional overhead right at the beginning. 🤔

**Comment by Unike267** [comment]
Hi!

> Could you share your assembly code of that loop? Maybe we can figure out why there is an additional overhead right at the beginning. 🤔

Yeah, of course! 

The waveform associated with the assembly code (first and second iteration):

![issue_cfs_7](https://github.com/stnolting/neorv32/assets/36209778/54d99a76-06ae-4f2b-bd4c-9c5a3111edc6)

*Note: use this waveform instead of the previous ones (I've rebuilt the code and it has changed a little bit).*

The assembly code of the first iteration:

``` asm
            NEORV32_CFS->REG[0] = fir; // Write fir to CFS memory-mapped register 0
 1c0:	00010737          	lui	a4,0x10
 1c4:	00170713          	addi	a4,a4,1 # 10001 <__neorv32_ram_size+0xe001>
 1c8:	b0e42023          	sw	a4,-1280(s0)
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fir to mult
 1cc:	00100693          	li	a3,1
 1d0:	b0d42223          	sw	a3,-1276(s0)
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fir from mult
 1d4:	00200713          	li	a4,2
 1d8:	b0e42223          	sw	a4,-1276(s0)
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
 1dc:	b0042223          	sw	zero,-1276(s0)
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
 1e0:	b0042603          	lw	a2,-1280(s0)
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
```

The assembly code of the second iteration:

``` asm
 1e4:	000015b7          	lui	a1,0x1
            NEORV32_CFS->REG[0] = sec; // Write sec to CFS memory-mapped register 0
 1e8:	00020637          	lui	a2,0x20
 1ec:	00260613          	addi	a2,a2,2 # 20002 <__neorv32_ram_size+0x1e002>
 1f0:	b0c42023          	sw	a2,-1280(s0)
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write sec to mult
 1f4:	b0d42223          	sw	a3,-1276(s0)
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read sec from mult
 1f8:	b0e42223          	sw	a4,-1276(s0)
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
 1fc:	b0042223          	sw	zero,-1276(s0)
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0  
 200:	b0042603          	lw	a2,-1280(s0)
```

*Note: the rest of the iterations are analogous to this one.*

As we can see the extra latencies are added in `pc: 1CC` and in `pc: 1D4` (*see the waveform*) with the instructions `ir: 0x00100693 ` and `ir: 0x00200713` respectively. These are two `load immediate` pseudo-instructions. I don't know why the compiler puts them in the first iteration and not in the rest of iterations 😅. 

I don't know if this tells you something. 😐

Anyway, thanks for your interest,

Cheers! 😃

**Comment by stnolting** [comment]
> I don't know if this tells you something. 

Maybe this is just the most efficient to emit code for the compiler. So it seems like this is just coding style issue. Instead of your individual if clauses you could try a large switch construct:

```c
switch(i) {
  case 0:  ... break;
  case 1:  ... break;
  case 3:  ... break;
  case 4:  ... break;
  default: ... break;
}
```

Perhaps the compiler can "better" optimize the individual iterations with this coding style.

**Comment by Unike267** [comment]
Hi @stnolting !

I've tried with the following code:

```c
while(i < 5)
    {
    // Write 4 inputs to mult and read the outputs from mult one by one
        switch (i){
            case 1:
            NEORV32_CFS->REG[0] = fir; // Write fir to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fir to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fir from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0          
            break;               
            case 2:
            NEORV32_CFS->REG[0] = sec; // Write sec to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write sec to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read sec from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0               
            break;
            case 3:
            NEORV32_CFS->REG[0] = thi; // Write thi to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write thi to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read thi from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0             
            break;                
            case 4:
            NEORV32_CFS->REG[0] = fou; // Write fou to CFS memory-mapped register 0
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fou to mult
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fou from mult
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0    
            break;
            default:
            // Intro
            neorv32_uart0_printf("\n CFS-lat \n\n");
            break;
            }
        i++;
    }
```

And I've obtained the following latencies for each iteration:

![issue_cfs_8](https://github.com/stnolting/neorv32/assets/36209778/11e5a11e-f7cd-424c-9c25-d39c20fdab3c)

It has changed from 24,20,20,20 to 22,20,20,24 😓

Waveform of the first and the second iteration:

![issue_cfs_9](https://github.com/stnolting/neorv32/assets/36209778/fe76fd68-9091-4f3f-b2b1-b10a030d4e8b)

Waveform of the third and the fourth iteration:

![issues_cfs_10](https://github.com/stnolting/neorv32/assets/36209778/40453b19-9ad9-485a-8504-8f6c1d732464)

In addition, it now uses more code between iterations. 😬

Viewing the assembly code the extra latencies is produced by `li` pseudo-instruction:

In `pc: 248` in the first iteration:

```asm
           NEORV32_CFS->REG[0] = fir; // Write fir to CFS memory-mapped register 0
 234:	00010637          	lui	a2,0x10
 238:	fffff737          	lui	a4,0xfffff
 23c:	00160613          	addi	a2,a2,1 # 10001 <__neorv32_ram_size+0xe001>
 240:	b0c72023          	sw	a2,-1280(a4) # ffffeb00 <__crt0_stack_end+0x7fffcb01>
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fir to mult
 244:	b0f72223          	sw	a5,-1276(a4)
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fir from mult
 248:	00200793          	li	a5,2
 24c:	b0f72223          	sw	a5,-1276(a4)
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
 250:	b0072223          	sw	zero,-1276(a4)
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0          
 254:	b0072783          	lw	a5,-1280(a4)
```

And in `pc: 2B4` and `pc: 2BC` in the fourth iteration.

```asm
 NEORV32_CFS->REG[0] = fou; // Write fou to CFS memory-mapped register 0
 2a4:	00080737          	lui	a4,0x80
 2a8:	fffff537          	lui	a0,0xfffff
 2ac:	00870713          	addi	a4,a4,8 # 80008 <__neorv32_ram_size+0x7e008>
 2b0:	b0e52023          	sw	a4,-1280(a0) # ffffeb00 <__crt0_stack_end+0x7fffcb01>
            NEORV32_CFS->REG[1] = 1; // Write 01 to CFS memory-mapped register 1; Write fou to mult
 2b4:	00100713          	li	a4,1
 2b8:	b0e52223          	sw	a4,-1276(a0)
            NEORV32_CFS->REG[1] = 2; // write 10 to CFS memory-mapped register 1; Read fou from mult
 2bc:	00200713          	li	a4,2
 2c0:	b0e52223          	sw	a4,-1276(a0)
            NEORV32_CFS->REG[1] = 0; // Write 00 to CFS memory-mapped register 1; Clean the control signals
 2c4:	b0052223          	sw	zero,-1276(a0)
            NEORV32_CFS->REG[0]; // Read mult result from CFS memory-mapped register 0    
 2c8:	b0052783          	lw	a5,-1280(a0)
```

*Note: this extra latency can also be seen in the waveforms.*

It seems that this issue resists us. 😆

Anyway, thank you very much for the hint! 😁

Cheers! 😃

**Comment by stnolting** [comment]
Oh man... 😅
One final thing you could try (before moving to inline assembly) is to change the compiler optimization level. The default is `-Os` - optimize for size. You could try "no optimization" (`-O0`) of "maximum optimization" (`-O3`). E.g.:

```bash
$ make MARCH=... EFFORT=-O3 clean_all ...
```

**Comment by Unike267** [comment]
Hi!

I've tried with the `EFFORT=-O0` and `EFFORT=-O3` options.

For `-O0` option I've obtained the following latencies for each iteration:

![issue_cfs_12](https://github.com/stnolting/neorv32/assets/36209778/17512d5a-4d42-478d-a41a-b5fde97110c4)

Waveform of the first iteration (`-O0`):

![issues_cfs_11](https://github.com/stnolting/neorv32/assets/36209778/3cd39778-1782-4799-b4ef-97b55eca5605)

For `-O3` option I've obtained the following latencies for each iteration:

![issue_cfs_13](https://github.com/stnolting/neorv32/assets/36209778/983d83e0-9f56-48bb-b25a-e778a8ea05c1)

Waveform of all iterations (`-O3`):

![issue_cfs_14](https://github.com/stnolting/neorv32/assets/36209778/56d525a6-6f03-483f-8f3f-75b07c6c56f7)

From these results we can conclude several things:

- The choice of the compiler optimization mode has a significant impact on the number of execution cycles.
- With the "no optimization" option we get the same number of cycles in all iterations but it takes twice as long. 
- With the "maximum optimization" option the cycles between iterations are minimized.

Finally, I think I will choose the 20 cycle measurement and I will indicate that it can fluctuate between 2 and 8 cycles depending on the compiler optimization and coding style. 

Thank you very much for your interest/hints. 😆 Really, I've learned a lot of very interesting things in this issue. 💪

Cheers! 😃
